File: src/main/java/me/lordsaad/modeoff/common/EventHandler.java
package me.lordsaad.modeoff.common;

import me.lordsaad.modeoff.api.PlotManager;
import net.minecraft.util.math.BlockPos;
import net.minecraftforge.event.entity.player.PlayerEvent;
import net.minecraftforge.event.entity.player.PlayerInteractEvent;
import net.minecraftforge.event.world.BlockEvent;
import net.minecraftforge.fml.common.eventhandler.Event;
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent;

/**
 * Created by LordSaad.
 */
public class EventHandler {

	@SubscribeEvent
	public void leftClickBlock(PlayerInteractEvent.LeftClickBlock event) {
		if (CommonProxy.teamMembers.contains(event.getEntityPlayer().getUniqueID())) return;
		if (!CommonProxy.contestants.contains(event.getEntityPlayer().getUniqueID())) {
			event.setUseItem(Event.Result.DENY);
			event.setUseBlock(Event.Result.DENY);
			event.setCanceled(true);
			return;
		}
		PlotManager manager = new PlotManager(event.getEntityPlayer());
		if (manager.plotID < 0) {
			event.setUseItem(Event.Result.DENY);
			event.setUseBlock(Event.Result.DENY);
			event.setCanceled(true);
			return;
		}
		if (manager.corner1 == null || manager.corner2 == null) {
			event.setUseItem(Event.Result.DENY);
			event.setUseBlock(Event.Result.DENY);
			event.setCanceled(true);
			return;
		}

		if (isWithinBounds(manager.corner1, manager.corner2, event.getPos())) event.setCanceled(true);
	}

	@SubscribeEvent
	public void onBreakBlock(BlockEvent.BreakEvent event) {
		if (CommonProxy.teamMembers.contains(event.getPlayer().getUniqueID())) return;
		if (!CommonProxy.contestants.contains(event.getPlayer().getUniqueID())) {
			event.setCanceled(true);
			return;
		}

		PlotManager manager = new PlotManager(event.getPlayer());
		if (manager.plotID < 0) {
			event.setCanceled(true);
			return;
		}
		if (manager.corner1 == null || manager.corner2 == null) {
			event.setCanceled(true);
			return;
		}

		if (isWithinBounds(manager.corner1, manager.corner2, event.getPos())) event.setCanceled(true);
	}

	@SubscribeEvent
	public void breakSpeed(PlayerEvent.BreakSpeed event) {
		if (CommonProxy.teamMembers.contains(event.getEntityPlayer().getUniqueID())) return;
		if (!CommonProxy.contestants.contains(event.getEntityPlayer().getUniqueID())) {
			event.setCanceled(true);
			return;
		}
		PlotManager manager = new PlotManager(event.getEntityPlayer());
		if (manager.plotID < 0) {
			event.setCanceled(true);
			return;
		}
		if (manager.corner1 == null || manager.corner2 == null) {
			event.setCanceled(true);
			return;
		}

		if (isWithinBounds(manager.corner1, manager.corner2, event.getPos())) event.setCanceled(true);
	}

	@SubscribeEvent
	public void place(BlockEvent.PlaceEvent event) {
		if (CommonProxy.teamMembers.contains(event.getPlayer().getUniqueID())) return;
		if (!CommonProxy.contestants.contains(event.getPlayer().getUniqueID())) {
			event.setCanceled(true);
			return;
		}

		PlotManager manager = new PlotManager(event.getPlayer());
		if (manager.plotID < 0) {
			event.setCanceled(true);
			return;
		}
		if (manager.corner1 == null || manager.corner2 == null) {
			event.setCanceled(true);
			return;
		}

		if (isWithinBounds(manager.corner1, manager.corner2, event.getPos())) event.setCanceled(true);
	}

	private boolean isWithinBounds(BlockPos corner1, BlockPos corner2, BlockPos pos) {
		int xMin = Math.min(corner1.getX(), corner2.getX());
		int xMax = Math.max(corner1.getX(), corner2.getX());
		int zMin = Math.min(corner1.getZ(), corner2.getZ());
		int zMax = Math.max(corner1.getZ(), corner2.getZ());

		return pos.getX() < xMin
				|| pos.getX() > xMax
				|| pos.getZ() < zMin
				|| pos.getZ() > zMax;
	}
}

File: src/main/kotlin/info/modoff/modeoff/common/EventHandler.kt
package info.modoff.modeoff.common

import info.modoff.modeoff.Modeoff
import info.modoff.modeoff.util.side
import net.minecraft.entity.player.EntityPlayerMP
import net.minecraft.util.math.BlockPos
import net.minecraftforge.event.entity.player.PlayerEvent
import net.minecraftforge.event.entity.player.PlayerInteractEvent
import net.minecraftforge.event.world.BlockEvent
import net.minecraftforge.fml.common.eventhandler.Event
import net.minecraftforge.fml.common.eventhandler.SubscribeEvent
import net.minecraftforge.fml.common.gameevent.PlayerEvent as FMLPlayerEvent

/**
 * Created by LordSaad.
 */
class EventHandler {
    @SubscribeEvent
    fun leftClickBlock(event: PlayerInteractEvent.LeftClickBlock) {
        val plotManager = Modeoff.proxy.getPlotManager(event.side)
        if (plotManager != null) {
            val plots = plotManager.getPlotsForPosition(event.pos)
            if (plots.isNotEmpty()) {
                event.useItem = Event.Result.DENY
                event.useBlock = Event.Result.DENY
                event.isCanceled = true
                return
            }
        }
    }

    @SubscribeEvent
    fun onBreakBlock(event: BlockEvent.BreakEvent) {
        val plotManager = Modeoff.proxy.getPlotManager(event.world.side)
        if (plotManager != null) {
            val plots = plotManager.getPlotsForPosition(event.pos)
            if (plots.isNotEmpty()) {
                event.isCanceled = true
                return
            }
        }
    }

    @SubscribeEvent
    fun breakSpeed(event: PlayerEvent.BreakSpeed) {
        val plotManager = Modeoff.proxy.getPlotManager(event.entity.world.side)
        if (plotManager != null) {
            val plots = plotManager.getPlotsForPosition(event.pos)
            if (plots.isNotEmpty()) {
                event.isCanceled = true
                return
            }
        }
    }

    @SubscribeEvent
    fun place(event: BlockEvent.PlaceEvent) {
        val plotManager = Modeoff.proxy.getPlotManager(event.player.world.side)
        if (plotManager != null) {
            val plots = plotManager.getPlotsForPosition(event.pos)
            if (plots.isNotEmpty()) {
                event.isCanceled = true
                return
            }
        }
    }

    @SubscribeEvent
    fun onPlayerLoggedIn(event: FMLPlayerEvent.PlayerLoggedInEvent) {
        Modeoff.plotManagerServer!!.sendToPlayer(event.player as EntityPlayerMP)
    }
}
